src/SessionManager.php

```diff
@@ -130,9 +130,15 @@ public function start($preserveStorage = false)
 
         session_start();
 
-        if ($oldSessionData instanceof \Traversable
-            || (! empty($oldSessionData) && is_array($oldSessionData))
-        ) {
+        // convert session data to plain array that’ll be acceptable as
+        // ArrayUtils::merge parameter
+        if ($oldSessionData instanceof Storage\StorageInterface) {
+            $oldSessionData = $oldSessionData->toArray();
+        } elseif ($oldSessionData instanceof \Traversable) {
+            $oldSessionData = iterator_to_array($oldSessionData);
+        }
+
+        if (! empty($oldSessionData) && is_array($oldSessionData)) {
             $_SESSION = ArrayUtils::merge($oldSessionData, $_SESSION, true);
```

Comment:
How it's going to work? As I can see `merge` has type hint on param 1 and 2 (`array`).
Your fix is in case when `$_SESSION` is NOT an array but above you pass (not an array) as the 2nd param?

Reply:
If I understand the code correctly, the new `$_SESSION` is passed to `merge()` after `session_start()` was called. Have a look at the [original code](https://github.com/zendframework/zend-session/blob/release-2.8.0/src/SessionManager.php#L126-L137) — it’s assumed (I guess so, because there’s no validation) that the new `$_SESSION` is something acceptable by `merge`. `session_start()` somehow modifies the `$_SESSION` superglobal.

I should move that `session_start()` call below my new code to make it easier to understand, shouldn’t I?